src/language_server_protocol.rs

```diff
@@ -3060,56 +3054,35 @@ impl LanguageClient {
                 .notify("s:AddHighlights", json!([source, highlights]))?;
         }
 
+        let bufnr = self.vim()?.get_bufnr(&filename, params)?;
+        self.draw_virtual_texts(&filename, viewport, bufnr)?;
```

Comment:
I feel this is getting more imperative than I'd imagine.

Ideally, I plan to move states into a standalone struct, such that when diagnostics, codelens etc, it internally update vim display to keep sync. No need to do in this PR though. Just for sharing ideas.